### PR TITLE
ENH: Add smooth cuboid phantom

### PR DESCRIPTION
**Pull request by adler-j**
Adds the following phantoms:

smooth_cuboid_x_2d-1
tgv_phantom

### CODE REVIEW

**Comment by adler-j** [comment]
Updated with the TGV phantom, good to go now

File: odl/phantom/geometric.py [hunk 1]
```diff
@@ -613,44 +614,191 @@ def ellipsoid_phantom(space, ellipsoids):
         raise ValueError('dimension not 2 or 3, no phantom available')
 
 
+def smooth_cuboid(space, min_pt=None, max_pt=None, axis=0):
+    """Rectangular cuboid with smooth variations.
```
Comment (kohr-h) [inline odl/phantom/geometric.py]: ~~Rectangular~~

File: odl/phantom/geometric.py [hunk 2]
```diff
@@ -613,44 +614,191 @@ def ellipsoid_phantom(space, ellipsoids):
...
+    ----------
+    space : `DiscreteLp`
+        Discretized space in which the phantom is supposed to be created.
+    min_pt : array-like of shape ``(space.ndim,)``, optional
+        Lower left corner of the cuboid. If ``None`` is given, a quarter
+        of the extent from ``space.min_pt`` towards the inside is chosen.
+    max_pt : array-like of shape ``(space.ndim,)``, optional
+        Upper right corner of the cuboid. If ``None`` is given, ``min_pt``
+        plus half the extent is chosen.
+    axis : `int` or sequence of `int`
```
Comment (kohr-h) [inline odl/phantom/geometric.py]: Remove ticks

File: odl/phantom/geometric.py [hunk 3]
```diff
@@ -613,44 +614,191 @@ def ellipsoid_phantom(space, ellipsoids):
...
+        Dimension(s) along which the smooth variation should happen.
+
+    Returns
+    -------
+    phantom : ``space``-element
+        The generated cuboid phantom in ``space``.
+    """
+    dom_min_pt = np.asarray(space.domain.min())
+    dom_max_pt = np.asarray(space.domain.max())
```
Comment (kohr-h) [inline odl/phantom/geometric.py]: Should be arrays already.

File: odl/phantom/geometric.py [hunk 4]
```diff
@@ -613,44 +614,191 @@ def ellipsoid_phantom(space, ellipsoids):
...
+    -----
+    The original phantom is given by a specific image. In this implementation,
+    we extracted the underlying parameters and the phantom thus works with
+    spaces of any shape. Due to this, small variations may occur when compared
+    to the original phantom.
+
+    References
+    ----------
+    [Bre+2010] K. Bredies, K. Kunisch, and T. Pock.
+    Total Generalized Variation. SIAM Journal on Imaging Sciences,
```
Comment (kohr-h) [inline odl/phantom/geometric.py]: `*Total Generalized Variation*.`

File: odl/phantom/geometric.py [hunk 5]
```diff
@@ -613,44 +614,191 @@ def ellipsoid_phantom(space, ellipsoids):
...
+        Needs to be two-dimensional.
+
+    Returns
+    -------
+    phantom : ``space``-element
+        The generated phantom in ``space``. Values have range [0, 1].
+
+    Notes
+    -----
+    The original phantom is given by a specific image. In this implementation,
```
Comment (kohr-h) [inline odl/phantom/geometric.py]: Maybe Kristian has code to generate the phantom for other scalings, too? Did you ask?

File: odl/phantom/geometric.py [hunk 6]
```diff
@@ -613,44 +614,191 @@ def ellipsoid_phantom(space, ellipsoids):
...
+    to the original phantom.
+
+    References
+    ----------
+    [Bre+2010] K. Bredies, K. Kunisch, and T. Pock.
+    Total Generalized Variation. SIAM Journal on Imaging Sciences,
+    3(3):492–526, Jan. 2010
+    """
+    # Note that all parameters used here were extracted by hand and may not fit
+    # exactly with the reference image.
```
Comment (kohr-h) [inline odl/phantom/geometric.py]: Already mentioned in Notes.

File: odl/phantom/geometric.py [hunk 7]
```diff
@@ -613,44 +614,191 @@ def ellipsoid_phantom(space, ellipsoids):
...
+    smooth_cuboid(space, axis=[0, 1]).show('smooth_cuboid x-y 2d')
+
+    # TGV phantom
+    tgv_phantom(space).show('tgv_phantom')
 
     # cuboid 3D
-    discr = odl.uniform_discr([-1, -1, -1], [1, 1, 1], [300, 300, 300])
-    cuboid(discr).show('cuboid 3d')
+    space = odl.uniform_discr([-1, -1, -1], [1, 1, 1], [300, 300, 300])
+    cuboid(space).show('cuboid 3d')
```
Comment (kohr-h) [inline odl/phantom/geometric.py]: Good change

File: odl/phantom/geometric.py (hunk 5 above)
Comment (adler-j) [inline odl/phantom/geometric.py]: Nope, but it aint open source to the best of my knowledge so didn't bother :/

**Comment by adler-j** [comment]
Fixed comments, merge after checks

**Comment by adler-j** [comment]
Also bumping @MatthiasJE since you were asking for this :)

**Comment by MatthiasJE** [comment]
Obviously, one should change the TGV examples accordingly before it gets forgotten.

**Comment by adler-j** [comment]
> Obviously, one should change the TGV examples accordingly before it gets forgotten.

Will do right now

**Pull request merged**